content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,711 @@
+writing them as mnemonics with assembly language, to abstractions with variables, labels,
+functions, and so on.
+
+
+[Functional Programming](https://en.wikipedia.org/wiki/Functional_programming)
+is one of the paradigms for managing complexity with its main idea of writing code as a set of
+functions and using those functions to pass values around instead imperative style of steps
+to perform to mutate some data.
+
+Imagine you want to create an argument parser that parses your arguments into a struct like this:
+
+```rust
+struct Args {
+    name: String,
+    answer: i32,
+}
+```
+
+
+If you want to write an argument parser purely imperatively, look at the horrors that entails,
+think about how to scale and maintain this:
+
+
+```rust
+fn parse() -> Args {
+	let name = None;
+	let answer = None;
+
+	let args = std::env::args();
+	while let (Some(arg), Some(value)) = (args.next(), args.next()){
+		if arg == "--name" {
+			name = Some(value.to_string());
+		}
+		if arg == "--answer" {
+			answer = Some(value.parse::<i32>.unwrap()),
+		}
+	};
+
+	Args {
+		name: name.expect("'--name' not given"),
+		answer: answer.expect("'--answer' not given"),
+	}
+}
+```
+
+What if you could use functional programming to just describe what you want:
+
+```rust
+fn parse() -> Args {
+    let name = Arg("--name").parse(|n| n.to_string())
+    let answer = Arg("--answer").parse(|a| a.parse::<i32>)
+    let Args = name
+      .zip(answer)
+      .map(|name, answer| Args { name, answer })
+      .run()
+      .unwrap()
+}
+```
+
+Now lets see how we get there.
+
+# Parsing command line options
+
+The main goal for this tutorial is to implement a basic library to parse command line options.
+If you used `cargo` you are probably familiar with expressions like this:
+
+```console
+$ cargo build --bin hello      # build a binary "hello"
+$ cargo check --package server # check if package "server" contains any issues
+```
+
+Parsing command line options in general is a wide problem and to make it more manageable this
+tutorial is going to look only at option arguments with a long name: `--bin hello` and
+`--package server` parts. This approach however can parse pretty much anything,
+[`bpaf`](https://crates.io/crates/bpaf) is a finished library that uses it.
+
+## Humble beginning
+
+Tutorial starts with some magical handwavey way of taking things from `std::env::args()` and placing
+them into variables and introduces a simple way of taking them out of the variables.
+
+```rust
+// a magical container with a value already inside
+struct Magic(&'static str);
+impl Magic {
+    // a way to get a value out of the container
+    fn run(self) -> String {
+        self.0.into()
+    }
+}
+
+#[test]
+fn sample() {
+    // --name Bob
+    let name = Magic("Bob");
+    let _result = name.run(); // "Bob"
+}
+```
+
+At this point parser can consume multiple string arguments using `Magic` and `Magic::run`.
```

Comment:
Yeah, I still don't see how `Magic` connects to "multiple arguments"...

maybe 

```suggestion
At this point the parser can represent single string arguments using `Magic` and extract them via `Magic::run`.
```